### PR TITLE
Allow JS multiple declarations of ctor properties

### LINKED ISSUE #9880: Duplicate identifier in JS class assigning to method in constructor
In a JavaScript file under `--allowJs`:
```js
class Test {
  constructor() {
    this.foo = null;
  }
  foo() {}
}
```
Produces errors
> a.js(3,5): error TS2300: Duplicate identifier 'foo'.
a.js(5,3): error TS2300: Duplicate identifier 'foo'.

### PR DESCRIPTION
**Pull request by sandersn**
Fixes #9880 

When a property is declared in the constructor and on the prototype of an ES6 class, the property's symbol is discarded in favour of the method's symbol. That because the usual use for this pattern is to bind an instance function: `this.m = this.m.bind(this)`. In this case the type you want really is the method's type.

### CODE REVIEW

**Comment by sandersn** [comment]
@RyanCavanaugh @weswigham mind taking a look?

File: tests/cases/conformance/salsa/multipleDeclarations.ts [hunk 1]
```diff
@@ -7,4 +7,19 @@ function C() {
 }
 C.prototype.m = function() {
     this.nothing();
-};
+}
+
```
Comment (RyanCavanaugh) [inline tests/cases/conformance/salsa/multipleDeclarations.ts]: Need a test that does this in the opposite order (with the constructor below the method declarations)

Comment (sandersn) [inline tests/cases/conformance/salsa/multipleDeclarations.ts]: Done

File: src/compiler/types.ts [hunk 2]
```diff
@@ -2137,6 +2137,7 @@ namespace ts {
         /* @internal */ exportSymbol?: Symbol;  // Exported symbol associated with this symbol
         /* @internal */ constEnumOnlyModule?: boolean; // True if module contains only const enums or other modules with only const enums
         /* @internal */ isReferenced?: boolean; // True if the symbol is referenced elsewhere
+        /* @internal */ isDiscardable?: boolean; // True if a Javascript class property can be overwritten by a method
```
Comment (yuit) [inline src/compiler/types.ts]: nit: could you name the property ? (e.g `isMethodOverwritable` etc.)

File: src/compiler/binder.ts [hunk 3]
```diff
@@ -327,33 +329,37 @@ namespace ts {
                 }
 
                 if (symbol.flags & excludes) {
-                    if (node.name) {
-                        node.name.parent = node;
+                    if (symbol.isDiscardable) {
```
Comment (yuit) [inline src/compiler/binder.ts]: nit: you may want to mention this change in the comment [above ](https://github.com/Microsoft/TypeScript/pull/10123/files#diff-7455dc48a52c4974a3e3671fcfb1dfb9R308)

Comment (sandersn) [inline src/compiler/binder.ts]: done

File: src/compiler/types.ts (hunk 2 above)
Comment (sandersn) [inline src/compiler/types.ts]: done. I ended up with `isReplaceableByMethod` but tell if you want me to change it.

**Comment by sandersn** [comment]
@weswigham I just got a test timeout failure on one of the linux workers with `workerCount=4`. :crying_cat_face:

**Comment by weswigham** [comment]
Grrr.... Being unable to predict starvation is a pain. For now, rerun the
build and I'll reduce worker count to 3.

**Comment by yuit** [comment]
I like the new name :smile_cat:

**Comment by mhegazy** [comment]
👍

**Pull request merged**